### PR TITLE
reset state and postcode if it's not valid for the current country

### LINKED ISSUE #46270: Changing between countries with states results in an error in the Checkout block
### Issue description

When editing a saved address and switching between two countries with states, e.g. Japan and Hungary an error is shown in the address form.

```
The provided state (JP02) is not valid. Must be one of: BK, BE, BA, BZ, BU, CS, FE, GS, HB, HE, JN, KE, NO, PE, SO, SZ, TO, VA, VE, ZA
The provided postcode / ZIP is not valid
```

This shows that server-side validation is still happening against the old country.

### Reproduction

0. Disable local pickup (I had trouble reproducing with Local Pickup enabled)
1. ~Be logged in.~ *
2. Add an item to your cart and go to the Checkout block.
3. Select Japan as your country. You may use this example address: 2 Chome-19-23 Aobadai, Meguro City, Tokyo 153-0042, Japan
4. Place the order.
5. Add another item to your cart and go to the Checkout block.
6. On the saved address card, edit your address and go immediately to the Country field. Change it to Hungary.
7. Observe the error.

1* This also works if you're logged out, you just have to fill a complete address then change country

#### Video

https://www.loom.com/share/5b8faeac82d241feab166862046501e4

#### Screenshot

Screenshot 2024-04-06 at 15 17 12

### PR DESCRIPTION
**Pull request by senadir**
When switching countries, Checkout would sometimes clear out state and zipcode, there are some cases in which those fields are not cleared correctly, and the server end up validating the current country against the old state and postcode. This PR resets them when changing country, and make sure to reset them as well if you're changing country, this is because state and postcode would always be validated client side first.

Initially I tried fixing this at server side, but it caused issues when placing an order, this feels like a good middleground.

Closes https://github.com/woocommerce/woocommerce/issues/46270

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Select a country with states, like Algeria, and fill out a fill valid address: Oran, Oran, 31000.
2. Switch to a country without states, like Afghanistan.
3. Switch back to a country with states like Spain, notice lack of error.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Correctly clear out state and postcode when switching countries.

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/StoreApi/Routes/V1/CartUpdateCustomer.php
Comment (senadir) [inline plugins/woocommerce/src/StoreApi/Routes/V1/CartUpdateCustomer.php:257]: This is no longer needed now that we reset at sanization level, I also tested this PR steps and they're still working

https://github.com/woocommerce/woocommerce-blocks/pull/8460

File: plugins/woocommerce/src/StoreApi/Schemas/V1/AbstractAddressSchema.php
Comment (senadir) [inline plugins/woocommerce/src/StoreApi/Schemas/V1/AbstractAddressSchema.php]: We need to reset both postcode and state

**Review by nielslange — APPROVED** [review]
Thanks for working on this, @senadir. The fix works as expected. I just closed and re-opened the PR so that the CI checks can run. Let's ensure that all CI checks are passing before merging this PR.

**Comment by senadir** [comment]
Some tests are failing and they seem to be valid, will look into them.

**Review by opr — DISMISSED** [review]
When testing this using Store API, I was able to place an order with a `JP` address with an invalid postcode and an invalid state.

I think this is because we're unsetting invalid values in the `AbstractAddressSchema` `sanitize_callback` and this also happens on checkout. If there's a way we can limit that logic to run only on the `cart/update-customer` route that would be better.

**Comment by senadir** [comment]
I will look into that, it's strange because I didn't remove any validation (not counting the one in get_address_x), I added more validation, I will look into this. I knew this issue would be tricky.

**Comment by senadir** [comment]
I changed the approach in the PR, would love a second review.

**Review by opr — APPROVED** [review]
This is working nicely for me now, thanks Nadir!

**Pull request merged**
